linux/src/socket_transport_linux.c

```diff
@@ -0,0 +1,696 @@
+                        LogError("Socket connected (%" PRI_SOCKET ") from %s:%d", accepted_socket, hostname_addr, cli_addr.sin_port);
+
+                        // Create the socket handle
+                        // Codes_SOCKET_TRANSPORT_LINUX_11_075: [ socket_transport_accept shall allocate a SOCKET_TRANSPORT for the incoming connection and call sm_create and sm_open on the connection. ]
+                        result = malloc(sizeof(SOCKET_TRANSPORT));
+                        if (result == NULL)
+                        {
+                            LogError("failure allocating SOCKET_TRANSPORT: %zu", sizeof(SOCKET_TRANSPORT));
+                        }
+                        else
+                        {
+                            result->sm = sm_create("Socket_transport_win32");
+                            if (result->sm == NULL)
+                            {
+                                LogError("Failed calling sm_create in accept, closing incoming socket.");
+                            }
+                            else
+                            {
+                                SM_RESULT open_result = sm_open_begin(result->sm);
+                                if (open_result == SM_EXEC_GRANTED)
+                                {
+                                    // Codes_SOCKET_TRANSPORT_LINUX_11_076: [ If successful socket_transport_accept shall return the allocated SOCKET_TRANSPORT of type SOCKET_DATA. ]
+                                    result->type = SOCKET_CLIENT;
+                                    result->socket = accepted_socket;
+                                    sm_open_end(result->sm, true);
+                                    sm_exec_end(socket_transport->sm);
+                                    goto all_ok;
+                                }
+                                else
+                                {
+                                    LogError("sm_open_begin failed with %" PRI_MU_ENUM " in accept, closing incoming socket.", MU_ENUM_VALUE(SM_RESULT, open_result));
+                                }
+                                sm_destroy(result->sm);
+                            }
+                            // Codes_SOCKET_TRANSPORT_LINUX_11_077: [ If any failure is encountered, socket_transport_accept shall fail and return NULL. ]
+                            free(result);
+                            result = NULL;
+                        }
+                    }
+                    close(accepted_socket);
+                }
+                // Codes_SOCKET_TRANSPORT_LINUX_11_078: [ socket_transport_accept shall call sm_exec_end. ]
+                sm_exec_end(socket_transport->sm);
+            }
+        }
+        result = NULL;
+    }
+all_ok:
+    return result;
+}
+
+SOCKET_HANDLE socket_transport_get_underlying_socket(SOCKET_TRANSPORT_HANDLE socket_transport)
```

Comment:
I have to question why is this API needed...

Reply:
It's better if the underlying socket is not exposed - so there should be a very strong case for why it is needed

Reply:
You have to have it exposed for epoll on linux and Overlapped on Windows.  They both require the raw socket.